File: E. Functional Programming in Python/Luhn1.hs

-- Implementation of Credit Card Number Validation --
-- Follows guide at http://ozark.hendrix.edu/~yorgey/490/static/Haskell-intro.pdf --

-- toDigits works using point free notation. The unnamed argument will get passed
-- to show, creating a String from the Integer. Map will cause each character in that
-- String to be made as the head of an empty list which will be then passed to
-- read, which will make that list into an Integer.
toDigits :: Integer -> [Integer]
toDigits = map (read . (:[])) . show

-- toDigitsRev works the same way as toDigits, it simply reverses the list
toDigitsRev :: Integer -> [Integer]
toDigitsRev = reverse . map (read . (:[])) . show

-- doubleEveryOther utilizes a cycle and the zipWith function. zipWith is applied
-- to the reversed list and a cycle of 1 and 2. The first element in the list is
-- multiplied by 1 and the next by 2 and the next by 1 and so on until the list
-- is exhausted. The list is reversed in order to work from right to left.
doubleEveryOther :: [Integer] -> [Integer]
doubleEveryOther list = zipWith (*) (reverse list) (cycle [1,2])

-- sumDigits uses the map function to apply toDigits to all elements in list. This
-- creates lists of lists so concat is used to get a list of single nonlist elements.
-- The sum of this list is then returned.
sumDigits :: [Integer] -> Integer
sumDigits list = sum $ concat $ map toDigits list

-- validate takes a potential cc number, converts it to digits, doubles every other
-- digit from right to left, sums the digits and finally checks if it is divisible
-- by 10.
validate :: Integer -> Bool
validate number = (sumDigits $ doubleEveryOther $ toDigits number) `mod` 10 == 0

File: E. Functional Programming in Python/Luhn1.py
from pyrsistent import pvector, v
from toolz import interleave, accumulate, last, concat
from operator import add
from pyrsistent.typing import PVector

# Maps every "character" in the string to a function that returns an int,
# and creates a PVector (immutable list) out of the resulting ints
def toDigits(n: int) -> PVector[int]:
    return pvector(map(lambda c: int(c), str(n)))

# Interleave joins a sequence of sequences by alternating between the two. Here,
# interleave is called on a list of two lists. The first list, ints[::-1][0::2],
# is made of the even indeces of the reversed "ints" list. The second is made of
# the odd indeces of the reversed "ints" list, mapped so that they are doubled,
# and made into a PVector. A single PVector is made from the interleavened sequences,
# as interleave returns a lazy iterator.
def doubleEveryOther(ints: PVector[int]) -> PVector[int]:
    return pvector(interleave([ints[::-1][0::2],
        pvector(map(lambda x: x * 2, ints[::-1][1::2]))]))

# Concat reduces a list of lists of ints into a single list.
# Accumulate applies the add function such that the first element is added to the
# second, that value is added to the third, etc, and last returns the final sum
# of all vaues in the list.
def sumDigits(ints: PVector[int]) -> int:
    return last(accumulate(add, concat(map(lambda c: toDigits(c), ints))))

# isValid converts the potential credit card number to digits, doubles every other
# starting from the right, sums them up, and checks if that is divisible by 10
def isValid(int: int) -> bool:
    return sumDigits(doubleEveryOther(toDigits(int))) % 10 == 0

# An immutable PVector is created from mapping the isValid function to the list
# of potential credit card numbers.
def testCC() -> PVector[bool]:
    print(pvector(map(isValid, [1234567890123456, 1234567890123452])))

testCC()
